text/0000-tracked-properties.md

```diff
@@ -0,0 +1,1029 @@
+
+  ```js
+  import { computed } from '@ember/object';
+  import { tracked } from '@glimmer/tracking';
+
+  class Person {
+    firstName;
+    lastName;
+
+    @tracked middleName;
+
+    @computed('firstName', 'lastName')
+    get fullName() {
+      return `${this.firstName} ${this.middleName} ${this.lastName}`;
+    }
+  }
+
+  let person = new Person();
+
+  set(person, 'firstName', 'Tom');
+  set(person, 'lastName', 'Dale');
+
+  person.middleName = 'Tomster';
+
+  person.fullName; // 'Tom Tomster Dale'
+  ```
+
+It will still be required to use `set` when updating computed properties and
+their dependencies. In the future, this restriction could possibly be relaxed.
+
+#### `get` and `set`
+
+It is common in the classic model to set and consume plain object properties
+which are not computed properties, or in any other way special. Ember's `get`
+and `set` functions historically allowed this by giving us the ability to
+intercept all property changes and watch for mutations.
+
+This presents a problem for tracked properties, particularly because of the
+recent change in Ember to enable native Javascript getters to replace `get`.
+This change means that we have no way to intercept `get`, and consequently no
+way for tracked properties to know whether or not a plain property will later
+be updated with `set`.
+
+To demonstrate this case, consider the following service and component:
+
+```js
+const Config = Service.extend({
+  polling: {
+    shouldPoll: false,
+    pollInterval: -1,
+  },
+
+  init() {
+    this._super(...arguments);
+
+    fetch('config/api/url')
+      .then(r => r.json())
+      .then(polling => set(this, 'polling', polling));
+  },
+});
+```
+```js
+class SomeComponent extends Component {
+  @service config;
+
+  @tracked
+  get pollInterval() {
+    let { shouldPoll, pollInterval } = this.config.polling;
+
+    return shouldPoll ? pollInterval : -1;
+  }
+}
+```
+
+Let's walk through the flow here:
+
+1. The `SomeComponent` component is rendered for the first time, instantiating
+   the `Config` service (assuming this the first time it has ever been
+   accessed). The service's init hook kicks off an async request to get the
+   configuration from a remote URl.
+2. The tracked `pollInterval` property first accesses the service injection,
+   which is a computed property. The property is detected and added to the
+   tracked stack.
+3. We then access the plain, undecorated `polling` object. Because it is
+   is not tracked and not a computed property, tracked does not know that it
+   could update in the future.
+4. Sometime later, the async request returns with the configuration object. We
+   set it on the service, but because our tracked getter did not know this
+   property would update, it does not invalidate.
+
+In order to prevent this from happening, user's will have to use `get` when
+accessing any values which may be set with `set`, and are not computed
+properties.
```

Comment:
Maybe add a note here that the easy way out to avoid `get` and `set` in this situation is to refactor and mark `polling` as `@tracked`. It should also be highlighted that this then _requires_ that the `polling` property is re-set and not a nested key.

```js
const Config = Service.extend({
  polling: tracked({
    value: {
      shouldPoll: false,
      pollInterval: -1,
    }
  }),

  init() {
    this._super(...arguments);

    fetch('config/api/url')
      .then(r => r.json())
      .then(polling => this.polling = polling);
  },
});
```

Reply:
Can this be confirmed? Given:

```js

const Config = Service.extend({
  polling: tracked({
    value: {
      shouldPoll: false,
      pollInterval: -1,
    }
  }),

  init() {
    this._super(...arguments);

    fetch('config/api/url')
      .then(r => r.json())
      .then(polling => this.polling = polling);
  },
});
```

```js
class SomeComponent extends Component {
  @service config;
  @tracked
  get pollInterval() {
    let { shouldPoll, pollInterval } = this.config.polling;
     return shouldPoll ? pollInterval : -1;
  }
}
```

It has not always been true in the history of tracked properties that `pollInterval` is invalidated when `this.polling =` is called on the service. If this works it implies cross-object tracked properties.

@pzuraq is that intended here? Can be possibly be made more explicit in the RFC? I don't think there is an example in this doc for cross-object tracked properties, but maybe I missed it?

Reply:
@mixonic can you define what you mean by "cross-object tracked properties"? 

To be clear, what this example would be doing is installing the `polling` tracked property on the Config service. When the `polling` property is accessed by the component, _its_ tag is entangled with the `pollInterval` tracked property, so when `polling` is set, everything is invalidated. The individual properties on `polling` are _not_ tracked, so this would not work:

```js
const Config = Service.extend({
  polling: tracked({
    value: {
      shouldPoll: false,
      pollInterval: -1,
    }
  }),

  init() {
    this._super(...arguments);

    fetch('config/api/url')
      .then(r => r.json())
      .then(({ shouldPoll, pollInterval ) => {
        this.polling.shouldPoll = shouldPoll;
        this.polling.pollInterval = pollInterval;
      });
  },
});
```

Reply:
I think you have described what I mean! :-) There are two objects (the service and the component) and if a tracked property on one (the service) is accessed from a tracked property on the component then they are entangled. Updates to the property on the service will dirty the property on the component.

Do I have it right?

Reply:
Yep! If that's not clear we should definitely make it clear.

More specifically, _any_ tracked getter or property that is accessed from a tracked context (e.g. when an autotrack stack exists) will become entangled with the stack. Templates, tracked getters, and computed properties all start an autotrack stack I believe, so anything that is accessed while those are running will be pushed on.